Core/clim-basic/drawing/pattern.lisp

```diff
@@ -99,42 +99,50 @@ pattern, stencil, image etc)."))
   ((array :type (simple-array (unsigned-byte 32) 2)))
   (:documentation "Helper class of RGBA result of another pattern."))
 
+(declaim (ftype (function (t) (values (unsigned-byte 32) &optional nil))
+                %rgba-value))
 (defun %rgba-value (element)
   "Helper function collapsing uniform design into 4-byte RGBA value."
   (flet ((transform (parameter)
-           (logand (logand (truncate (* parameter 255)) 255))))
+           (logand (truncate (* parameter 255)) 255)))
     (etypecase element
       ((unsigned-byte 32) element)
       ;; Uniform-compositium is a masked-compositum rgb + opacity
       ((or color opacity uniform-compositum)
        (multiple-value-bind (red green blue opacity)
            (color-rgba element)
-         (logior (ash (transform red)   24)
-                 (ash (transform green) 16)
-                 (ash (transform blue)   8)
-                 (transform opacity))))
+         (logior (ash (transform red)    24)
+                 (ash (transform green)  16)
+                 (ash (transform blue)    8)
+                 (ash (transform opacity) 0))))
       (indirect-ink (%rgba-value (indirect-ink-ink element))))))
 
 (defgeneric %pattern-rgba-value (pattern x y)
   (:documentation "Returns a collapsed RGBA value for position [X, Y].")
   (:method ((pattern %rgba-pattern) (x fixnum) (y fixnum))
     (let ((array (pattern-array pattern)))
+      (declare (type (array (unsigned-byte 32) 2) array))
       (if (array-in-bounds-p array y x)
-          (aref (pattern-array pattern) y x)
+          (aref array y x)
           #x00000000)))
   (:method ((pattern indirect-ink) x y)
     (%rgba-value (design-ink pattern x y)))
-  (:method (design x y) ;; fallback method
+  (:method (design x y) ; fallback method
     (%rgba-value (design-ink design x y))))
 
 (defmethod design-ink ((pattern %rgba-pattern) x y)
   (let ((array (pattern-array pattern)))
+    (declare (type (array (unsigned-byte 32) 2) array))
     (if (array-in-bounds-p array y x)
-        (let ((elt (aref (pattern-array pattern) y x)))
-          (make-uniform-compositum (make-rgb-color (float (/ (ldb (byte 8 24) elt) 255))
-                                                   (float (/ (ldb (byte 8 16) elt) 255))
-                                                   (float (/ (ldb (byte 8 08) elt) 255)))
-                                   (float (/ (ldb (byte 8 00) elt) 255))))
+        (let* ((rgba-value (aref array y x))
+               (alpha (ldb (byte 8 0) rgba-value))
+               (color (make-rgb-color (/ (ldb (byte 8 24) rgba-value) 255.0)
+                                      (/ (ldb (byte 8 16) rgba-value) 255.0)
+                                      (/ (ldb (byte 8  8) rgba-value) 255.0))))
+          (case alpha
+            (0 +transparent-ink+)
+            (255 color)
+            (t (make-uniform-compositum color (/ alpha 255.0)))))
```

Comment:
alpha is known before the color is computed, so maybe

```
(let* ((rgba-value  …)
         (alpha …))
  (case alpha
    (0 +transaprent-ink+)
    (255 (make-color))
    (otherwise (make-uniform-compositum (make-color) (/ alpha 255.0))))
```
?

Reply:
another question: why do we avoid calling a function make-uniform-compositum? when alpha is opaque, then it returns the ink (but it accepts normalized alpha value).

```
(defun make-uniform-compositum (ink opacity-value)
  (cond ((= opacity-value 0)
         +transparent-ink+)
        ((= opacity-value 1)
         ink)
        (t
         (make-instance 'uniform-compositum
           :ink ink
           :mask (make-opacity opacity-value)))))
```

n.b since it is an internal function, we could define also `make-uniform-compositum*` which takes values [0,255] instead of [0,1] and call that for patterns.

Reply:
Not calling `make-uniform-compositum` avoids computing `(/ alpha 255.0)` for the (presumably common) cases which would throw it away immediately.

I'm not sure making `make-uniform-compositum*` is worth it. Especially if this is the only call site.

Reply:
Only computing the color if needed seems like a good idea.